cmd/axelard/cmd/vald/evm/evm.go

```diff
@@ -453,33 +453,74 @@ func parseTransferKeyConfirmationParams(cdc *codec.LegacyAmino, attributes map[s
 		nil
 }
 
-func (mgr Mgr) validate(rpc rpc.Client, txID common.Hash, confHeight uint64, validateTx func(tx *geth.Transaction, txReceipt *geth.Receipt) bool) bool {
+func getLatestFinalizedBlockNumber(rpc rpc.Client, confHeight uint64) (*big.Int, error) {
+	if rpc.IsMoonbeam() {
+		finalizedBlockHash, err := rpc.ChainGetFinalizedHead(context.Background())
+		if err != nil {
+			return nil, err
+		}
+
+		header, err := rpc.ChainGetHeader(context.Background(), finalizedBlockHash)
+		if err != nil {
+			return nil, err
+		}
+
+		return header.Number.ToInt(), nil
+	}
+
 	blockNumber, err := rpc.BlockNumber(context.Background())
 	if err != nil {
-		mgr.logger.Debug(sdkerrors.Wrap(err, "checking block number failed").Error())
-		// TODO: this error is not the caller's fault, so we should implement a retry here instead of voting against
-		return false
+		return nil, err
 	}
 
+	return big.NewInt(int64(blockNumber - confHeight + 1)), nil
+}
+
+func (mgr Mgr) validate(rpc rpc.Client, txID common.Hash, confHeight uint64, validateTx func(tx *geth.Transaction, txReceipt *geth.Receipt) bool) bool {
 	tx, _, err := rpc.TransactionByHash(context.Background(), txID)
 	if err != nil {
-		mgr.logger.Debug(sdkerrors.Wrap(err, "transaction by hash call failed").Error())
+		mgr.logger.Debug(sdkerrors.Wrap(err, "get transaction by hash call failed").Error())
 		return false
 	}
 
 	txReceipt, err := rpc.TransactionReceipt(context.Background(), txID)
 	if err != nil {
-		mgr.logger.Debug(sdkerrors.Wrap(err, "transaction receipt call failed").Error())
+		mgr.logger.Debug(sdkerrors.Wrap(err, "get transaction receipt call failed").Error())
 		return false
 	}
 
-	if !isTxFinalized(txReceipt, blockNumber, confHeight) {
-		mgr.logger.Debug(fmt.Sprintf("transaction %s does not have enough confirmations yet", txReceipt.TxHash.String()))
+	if !isTxSuccessful(txReceipt) {
+		mgr.logger.Debug(fmt.Sprintf("transaction %s failed", txReceipt.TxHash.String()))
 		return false
 	}
 
-	if !isTxSuccessful(txReceipt) {
-		mgr.logger.Debug(fmt.Sprintf("transaction %s failed", txReceipt.TxHash.String()))
+	latestFinalizedBlockNumber, err := getLatestFinalizedBlockNumber(rpc, confHeight)
+	if err != nil {
+		mgr.logger.Debug(sdkerrors.Wrap(err, "get latest finalized block number failed").Error())
+		return false
+	}
+
+	if latestFinalizedBlockNumber.Cmp(txReceipt.BlockNumber) < 0 {
+		mgr.logger.Debug(fmt.Sprintf("transaction %s is not finalized yet", txReceipt.TxHash.String()))
+		return false
+	}
+
+	txBlock, err := rpc.BlockByNumber(context.Background(), txReceipt.BlockNumber)
```

Comment:
lines 508-523 implement step 4 from here, right? [Consensus & Finality | Moonbeam Docs](https://docs.moonbeam.network/builders/get-started/eth-compare/consensus-finality/)

> As as safety check, retrieve the block by number, and verify that the given transaction hash is in the block

This is a linear search through all txs in a block.  Are we sure we want to do this for all evm chains?

Reply:
Yes, even for evm chains running proof-of-work, this is beneficial in terms of safety. Linear search isn't too bad considering how many evm txs usually fit one block.